### ISSUE TITLE
monitor-uri does not work with HTTP/2

### ISSUE THREAD

**Issue by TimWolla** [haproxy/haproxy#509]
## Output of `haproxy -vv` and `uname -a`

```
Linux *snip* 4.4.0-173-generic #203-Ubuntu SMP Wed Jan 15 02:55:01 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```
```
HA-Proxy version 2.2-dev2-e35d1d4-16 2020/02/11 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits
  OPTIONS = USE_OPENSSL=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL -LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 -ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with OpenSSL version : OpenSSL 1.0.2g  1 Mar 2016
Running on OpenSSL version : OpenSSL 1.0.2g  1 Mar 2016
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built without PCRE or PCRE2 support (using libc's regex instead)
Encrypted password support via crypt(3): yes
Built without compression support (neither USE_ZLIB nor USE_SLZ are set).
Compression algorithms supported : identity("identity")

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : none

Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[TRACE] trace
	[COMP] compression
```

Also seen with `2.1.3-1~bpo10+1` in production.

## What's the configuration?

```
frontend fe_http
	mode http

	bind *:8080 ssl crt ssl/haproxy.pem alpn h2,http/1.1

	monitor-uri /health

	use_backend be_http

backend be_http
	mode http

	server example example.com:80
```

## Steps to reproduce the behavior

1. `nghttp -v https://localhost:8080/health`
2. `curl -k -v https://localhost:8080/health`

## Actual behavior

1. Returns a 404 from the backend server.
2. Returns the 200 OK from `monitor-uri`.

## Expected behavior

I expected (1) to behave as (2).

## Do you have any idea what may have caused this?

I assume that monitor-uri is not handled in the HTTP/2 path.

## Do you have an idea how to solve the issue?

No.

**Comment by capflam** [comment]
You're right. The matching against the monitor-uri is done using the request uri, and not the request path. In HTTP2, the uri is absolute. I'll fix it.

**Comment by capflam** [comment]
Ok, it should be fixed now. But it is not really a bug, because it is stated in the documentation the matching is performed against the request's URI. not its path. In HTTP/2, most of time, clients send absolute-form URIs. But I agree it may be surprising.

**Comment by TimWolla** [comment]
> But it is not really a bug, because it is stated in the documentation the matching is performed against the request's URI. not its path.

However I would not be able to match both H1 and H2 _without_ this fix, because I only have one `monitor-uri` that I either can configure in absolute form or not, no?

In any case: Thanks for the patch! :smile:

**Comment by capflam** [comment]
True, it was the main motivation to push this patch. I will see with Willy if we backport it or not.

**Comment by haproxy-mirror** [comment]
On Tue, Feb 18, 2020 at 07:43:22AM -0800, Christopher Faulet wrote:
> True, it was the main motivation to push this patch. I will see with Willy if we backport it or not.

I'm not against a backport, the point being that the patch itself proves
that nobody ever managed to get monitor-uri to work with H2. This further
shows how outdated monitor-uri is. With the new return directive we'll
finally have much cleaner and flexible options in the future.

**Comment by TimWolla** [comment]
> With the new return directive we'll finally have much cleaner and flexible options in the future.

The thing I enjoy about `monitor-uri` is the automated suppression of the logs. In fact before discovering `monitor-uri` I implemented the health check response using Lua.

**Comment by haproxy-mirror** [comment]
On Tue, Feb 18, 2020 at 09:59:19AM -0800, Tim Düsterhus wrote:
> The thing I enjoy about `monitor-uri` is the automated suppression of the
> logs.

I know, we discussed with Christopher about having an extra "log-level"
directive on "return" to achieve the same :-) It could even be useful to
alert on certain conditions by the way.

**Comment by capflam** [comment]
FYI, the fix was backported to 2.1.
